src/coreComponents/common/LifoStorage.hpp

```diff
@@ -0,0 +1,198 @@
+
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "common/TimingMacros.hpp"
+#include "common/LifoStorageCommon.hpp"
+#include "common/LifoStorageHost.hpp"
+#ifdef GEOSX_USE_CUDA
+#include "common/LifoStorageCuda.hpp"
+#endif
+
+namespace geosx
+{
+/**
+ * This class is used to store in a LIFO way buffers, first on device, then on host, then on disk.
+ */
+template< typename T, typename INDEX_TYPE >
+class LifoStorage
+{
+
+public:
+
+
+  /**
+   * A LIFO storage will store numberOfBuffersToStoreDevice buffer on
+   * deevice, numberOfBuffersToStoreHost on host and the rest on disk.
+   *
+   * @param name                           Prefix of the files used to save the occurenncy of the saved buffer on disk.
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnDevice Maximum number of array to store on device memory. If negative opposite of the percent of left
+   * memory we want to use( -80 = use 80% of remaining memory ).
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory . If negative opposite of the percent of left
+   * memory we want to use( -80 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorage( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnDevice, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    m_maxNumberOfBuffers( maxNumberOfBuffers ),
+    m_bufferSize( elemCnt*sizeof( T ) ),
+    m_bufferCount( 0 )
+  {
+    LIFO_LOG_RANK( " LIFO : maximum size "<< m_maxNumberOfBuffers << " buffers " );
+    double bufferSize = ( ( double ) m_bufferSize ) / ( 1024.0 * 1024.0 );
+    LIFO_LOG_RANK( " LIFO : buffer size "<< bufferSize << "MB" );
+    if( numberOfBuffersToStoreOnDevice < 0 )
+    {
+#ifdef GEOSX_USE_CUDA
+      numberOfBuffersToStoreOnDevice = LifoStorageCuda< T, INDEX_TYPE >::computeNumberOfBufferOnDevice( -numberOfBuffersToStoreOnDevice, m_bufferSize, m_maxNumberOfBuffers );
+#else
+      numberOfBuffersToStoreOnDevice = 0;
+#endif
+    }
+    if( numberOfBuffersToStoreOnHost < 0 )
+    {
+      GEOSX_ERROR_IF( numberOfBuffersToStoreOnHost < -100, "Error, numberOfBuffersToStoreOnHost should be greater than -100" );
+      size_t free = sysconf( _SC_AVPHYS_PAGES ) * sysconf( _SC_PAGESIZE );
+      numberOfBuffersToStoreOnHost = std::max( 1, std::min( ( int )( -0.01 * numberOfBuffersToStoreOnHost * free / m_bufferSize ), m_maxNumberOfBuffers - numberOfBuffersToStoreOnDevice ) );
+      double freeGB = ( ( double ) free ) / ( 1024.0 * 1024.0 * 1024.0 ) / MpiWrapper::nodeCommSize();
+      LIFO_LOG_RANK( " LIFO : available memory on host " << freeGB << " GB" );
+    }
+    LIFO_LOG_RANK( " LIFO : allocating "<< numberOfBuffersToStoreOnHost <<" buffers on host" );
+    LIFO_LOG_RANK( " LIFO : allocating "<< numberOfBuffersToStoreOnDevice <<" buffers on device" );
```

Comment:
So please try to move this to `LifoStorageCommon` and delegate the code between the `ifdef/else/endif` to the appropriate constructors in `LifoStorageHost` and `LifoStorageCuda`.
That way you should be able to remove the `LifoStorage` class and use the base class `LifoStorageCommon` as base class (that you'll be able to rename `LifoStorage` I think.)

Reply:
I tried it this morning but I'm still stick with the choose to tak a Host or Cuda Lifo depending the values of the constructor parameters and the remaining memory at the given time. This choice could be moved to the user code maybe but still with ifdef etc...